src/util/emitUtils.js

```diff
@@ -560,13 +560,15 @@ function getNewObjAsString(key, obj) {
  */
 function emitToClients(io, nsp, rooms, key, obj) {
   const namespace = io.of(nsp);
-  if (rooms && rooms.length) {
+  if (rooms && rooms.length && rooms[0]) {
```

Comment:
why is rooms[0] needed?

Reply:
On the description ```Added a check in the first if statement of emitToClients as there was an edge case where the rooms array was [null] for IMC events.```

Reply:
Yeah, whenever events come from the refocus room itself (e.g activation/deactivation or user enter / leave ) it results in an array that looks like this `room = [ null ]`

This passes the if statement since `room` exists and `room.length` is 1 even though it contains no valid values. 

There might be a more valid place to check for this before this point.

Reply:
Add test for this?

Reply:
If we have to go with this solution of emitting to all bots for "bot.event" events that originate from rooms, it would be better to do it above at emitter.js:35 by passing in the list of all connected bots: 
`const bots = obj.botId ? [obj.botId] : u.connectedRooms['/bots'];`

But it would be even better to be able to filter the event and send it only to the clients that need it, like we do for everything else.

When a room initiates a "bot.event" event, does that event have to go to all bots? 
If so, why? Why not just the bots that are associated with the room? Is there any way around this?
If not, can we do the same thing as at emitter.js:40 and get all the associated bots to pass in to emitToClient?

Reply:
- When a room initiates a "bot.event" event, does that event have to go to all bots?

No, the event only really needs to go to the bots in the room, however in 9 out of 10 instances  all of the bots will be in a given room anyway.
Unfortunately emitting to particular bots in a given room would require quite a bit of work, currently `emitter.js:35` returns undefined for bots.

I think this should definitely be a future work item, but for the purposes of this story I think it's out of scope.

Reply:
Please create user story and let's make sure it gets prioritized.

Reply:
> Unfortunately emitting to particular bots in a given room would require quite a bit of work

I believe the only change required on the api side would be to attach the room along with its bots before publishing the event. I added something similar for room events as part of the original implementation (model/room.js:attachBotsAndPublish), and would have done the same for Events if I had realized they could be published without a botId.

And on the realtime side, the only change would be what I suggested above: 
"If not, can we do the same thing as at emitter.js:40 and get all the associated bots to pass in to emitToClient?"
Again, this is something that is already being done for room events, and would have been included originally had I been aware of this case.

Reply:
To clarify, I'm arguing for including these changes as part of this work. (or separately if you like, as long as it gets merged before switching over to the realtime app for IMC events)

I think fixing this in a way that's consistent with the original intent and the behavior in other cases is within the scope of "final fixes to get IMC events working".

Unless I'm missing something and the above changes wouldn't be sufficient?

Reply:
@ConalCosgrove

Reply:
Got this now, as I said, the line `emitter.js:40` was returning undefined as the value for bots each time. Turns out the code wasn't destructuring the object properly. 
Updated this code now - from `obj.type.bots` `obj.RoomType.Bots` - and the code is working.

Will work on a PR for refocus to add this functionality to events too.

Reply:
Ah. This is from the change in https://github.com/salesforce/refocus/pull/1192. I didn't realize it would change the attribute name on the room instance and require updating the realtime side.
Good catch.